### PR TITLE
Fix edge cases that prevent completion

### PR DESCRIPTION
**Pull request by faldor20**
I encountered a few cases where completion doesn't work correctly. Most obviously:
```ocaml
let a=[1]|>Li
```
This will not complete properly.
This was because the position_prefix code had a bunch of edge cases it didn't account for. I have added a whole collection of tests for these edge cases, and included two different proposed replacements for this prefix code, plus a benchmark of the difference between the old and new solutions.

I'm quite new to ocaml overall so i wanted to just run this by you all to get some feedback before i clean it up and settle on a solution.

After mucking around with this for many hours I think that the best solution would be to actually run the lexer over the tiny snippet and then use that token information to make an informed decision about what to feed back to merlin.

One key thing that occured to me is that currently none of these solution covers very odd edge cases like:
```ocaml
[1]|>List.
ma
[1]|> List. ma
```
but use lexer would solve this problem. though I'm honestly not sure if merlin would even deal with such a thing.

### CODE REVIEW

**Comment by faldor20** [comment]
```
┌─────────────────┬────────────┬─────────┬────────────┐
│ Name            │   Time/Run │ mWd/Run │ Percentage │
├─────────────────┼────────────┼─────────┼────────────┤
│ regex           │   916.98ns │  90.00w │     44.62% │
│ regex_long      │ 1_168.30ns │  97.00w │     56.85% │
│ regex_only      │   392.91ns │  51.00w │     19.12% │
│ regex_only_long │ 2_055.10ns │  50.99w │    100.00% │
│ old             │   269.50ns │  22.00w │     13.11% │
│ old_long        │   260.33ns │  22.00w │     12.67% │
└─────────────────┴────────────┴─────────┴────────────┘
```
these are the results of running it on my laptop. 
This could get a bit if we are happy to reverse the list still.  but it does make the regex pretty weird, because it's backwards

**Comment by rgrinberg** [comment]
I'm curious for why you added the benchmarks, anything in particular you expect to be slow here?

File: ocaml-lsp-server/src/compl.ml [hunk 1]
```diff
@@ -95,8 +103,51 @@ let prefix_of_position ~short_path source position =
...
     else reconstructed_prefix
 
-(** [suffix_of_position source position] computes the suffix of the identifier
-    after [position]. *)
+module String = struct
+  include String
+
+  (**Filters a string keeping any chars for which f returns true and discarding
+     those for which it returns false*)
+  let filter str ~f =
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/compl.ml:111]: General function should be moved to import.ml

File: ocaml-lsp-server/src/compl.ml [hunk 2]
```diff
@@ -260,8 +311,12 @@ module Complete_with_construct = struct
       List.mapi constructed_exprs ~f:completionItem_of_constructed_expr
 end
 
+let logCompletion log =
+  Log.log ~section:"resolveCompletion" (fun () -> Log.msg log [])
+
 let complete (state : State.t)
     ({ textDocument = { uri }; position = pos; _ } : CompletionParams.t) =
+  logCompletion "ho1";
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/compl.ml:319]: junk?

File: ocaml-lsp-server/src/compl.ml [hunk 3]
```diff
@@ -365,10 +426,12 @@ let format_doc ~markdown doc =
 
 let resolve doc (compl : CompletionItem.t) (resolve : Resolve.t) query_doc
     ~markdown =
+  logCompletion "Starting  completion";
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/compl.ml:429]: Why the extra space?

File: ocaml-lsp-server/src/compl.ml [hunk 4]
```diff
@@ -380,6 +443,7 @@ let resolve doc (compl : CompletionItem.t) (resolve : Resolve.t) query_doc
                 (Document.Merlin.source doc)
                 logical_position
             in
+            logCompletion @@ "completion prefix is:" ^ prefix;
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/compl.ml:446]: This log message is allocating its argument if no logging is enabled. Defeating the purpose of having `Log.log` taking a function argument.

File: ocaml-lsp-server/src/compl.ml [hunk 5]
```diff
@@ -392,7 +456,9 @@ let resolve doc (compl : CompletionItem.t) (resolve : Resolve.t) query_doc
               character = position.character + String.length suffix
             }
           in
+
           let range = Range.create ~start ~end_ in
+
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/compl.ml:461]: pointless space

File: ocaml-lsp-server/src/compl.mli [hunk 6]
```diff
@@ -27,7 +27,9 @@ val resolve :
   -> CompletionItem.t Fiber.t
 
 (** [prefix_of_position ~short_path source position] computes prefix before
-    given [position].
+    given [position]. 
+    A prefix is essentially a piece of code that refers to one thing eg a single infix operator "|>", a single reference to a function or variable: "List.map" a keyword "let" etc
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/compl.mli:31]: 80 chars wrapping

File: ocaml-lsp-server/bench/ocaml_lsp_bench.ml [hunk 7]
```diff
@@ -0,0 +1,44 @@
+open Core
+open Core_bench
+open Ocaml_lsp_server.Testing
```
Comment (rgrinberg) [inline ocaml-lsp-server/bench/ocaml_lsp_bench.ml:3]: Don't open this module and just qualify everything you need as `Testing.` please.

File: ocaml-lsp-server/src/prefix_parser.ml [hunk 8]
```diff
@@ -0,0 +1,31 @@
+open Re
+
+(*Regex based parser*)
+
+let name_or_label_regex =
+  Re.compile @@ Re.Posix.re {|([~?`]|let%|and%)?([a-zA-Z0-9_']|[a-zA-Z0-9_'] *\. *)+$|}
+
+(** matches let%lwt and let* style expressions. See here:https://v2.ocaml.org/manual/bindingops.html *)
+let monadic_bind =
+  Re.compile @@ Re.Posix.re {|(let|and)([$&*+\-/=>@^|.]|(%[a-zA-Z0-9_']*))$|}
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/prefix_parser.ml:10]: Can you use the combinators from `Re` instead of `Re.Posix` please.

**Comment by faldor20** [comment]
@rgrinberg Well, know not particularly. But for code that runs almost every keypress I think speed is key. I wanted to validate that I wasn't causing any meaningful slowdown

File: ocaml-lsp-server/src/compl.ml (hunk 4 above)
Comment (faldor20) [reply ocaml-lsp-server/src/compl.ml:446]: Yeah, good point

File: ocaml-lsp-server/src/compl.ml (hunk 5 above)
Comment (faldor20) [reply ocaml-lsp-server/src/compl.ml:461]: I thought ocamlfmt would have dealt with those, oh well done now :)

Comment (faldor20) [reply ocaml-lsp-server/src/compl.ml:461]: I thought ocamlfmt would have dealt with those, oh well done now :)

File: ocaml-lsp-server/src/prefix_parser.ml (hunk 8 above)
Comment (faldor20) [reply ocaml-lsp-server/src/prefix_parser.ml:10]: How set are you on this? I far far prefer to use standard regex because 
a) it's something universal
b) it means it's easy to test and debug using tools like https://regexr.com/ 
But if you think there are good reasons to use the combinators I can convert it

File: ocaml-lsp-server/src/prefix_parser.ml [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+let infixRegex = Re.compile @@ Re.Posix.re {|[~?:!$&*+\-\/=><@^|%<.#]+$|}
+
+open Import
+
+module Option = struct
+  include Option
+
+  let none_bind func option =
```
[the remark this answers is not in the document]
Comment (faldor20) [reply ocaml-lsp-server/src/prefix_parser.ml:24]: do you mean make multiple copies of that some logic? or just to use the full match logic in every case?

File: ocaml-lsp-server/test/e2e/__tests__/textDocument-completion.test.ts [hunk 10]
```diff
@@ -91,6 +91,395 @@ describe_opt("textDocument/completion", () => {
     expect(items).toMatchObject([{ label: "StringLabels" }]);
   });
 
+  it("can start completion after operator with space", async () => {
```
[the remark this answers is not in the document]
Comment (faldor20) [reply ocaml-lsp-server/test/e2e/__tests__/textDocument-completion.test.ts:94]: I didn't really want to migrate the whole test suite in this PR and i didn't see any existing completion tests.

Would it be okay to just merge this with the new tests in the old suite and then bundle them in with a larger effort to convert all the completion tests to the new e2e suite?

**Comment by faldor20** [comment]
Also I've switched to the reversed implementation because its performance is more predictable and writing some regexes backwards isn't that onerous

**Comment by faldor20** [comment]
As for the benchmarks, I understand they may not make sense in this PR and you may want a different benchmarking lib etc. But I do think it makes sense to have something setup at some point. Even just for testing little things to ensure that new changes aren't causing perf regressions. I'm sure some people out there can always accurately guess the performance effect of a change just by looking at it, but I think microbenchmarks are a very good thing to have around.

An example of this is that i find this server mostly very performant(kudos, the startup time is super impressive) but every now and again it gets into a weird state or i open a particular file that causes major slowdown. I'd love to find out why

File: ocaml-lsp-server/src/prefix_parser.ml
Comment (rgrinberg) [inline ocaml-lsp-server/src/prefix_parser.ml:1]: This module needs an mli

File: ocaml-lsp-server/src/prefix_parser.ml (hunk 8 above)
Comment (rgrinberg) [reply ocaml-lsp-server/src/prefix_parser.ml:13]: I think the main issue with the way you've written it is that you're using capturing groups that you don't actually need. That's a little confusing and also slows down the matching.

File: ocaml-lsp-server/src/prefix_parser.ml (hunk 9 above)
[the remark this answers is not in the document]
Comment (rgrinberg) [reply ocaml-lsp-server/src/prefix_parser.ml:24]: The explicit match would be better. I usually don't like introducing helper functions that nobody is familiar with. Put another way, my criteria is that it should be present in at least one popular stdlib before introducing it here.

You can also rewrite your code in the following way if I'm not mistaken:

```
List.find_map [ name_or_label_regex_rev ; monadic_bind_rev ; infixRegex_rev ] ~f:(fun re ->
  Re.exec_opt re text)
```

File: ocaml-lsp-server/test/e2e/__tests__/textDocument-completion.test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (rgrinberg) [reply ocaml-lsp-server/test/e2e/__tests__/textDocument-completion.test.ts:94]: If you don't have time, I can do it myself. I'm afraid that the larger effort to convert everything to the new test suite is never going to come. So for now, we're just not going to let the current test suite grow.

File: ocaml-lsp-server/src/prefix_parser.ml [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+let name_or_label_regex_rev =
+  Re.compile @@ Re.Posix.re {|^([a-zA-Z0-9_']| *\. *[a-zA-Z0-9_'])+([~?`]|%tel|%dna)?|}
+
+(** matches let%lwt and let* style expressions. See here:https://v2.ocaml.org/manual/bindingops.html *)
+let monadic_bind =
+  Re.compile @@ Re.Posix.re {|(let|and)([$&*+\-/=>@^|.]|(%[a-zA-Z0-9_']*))$|}
+let monadic_bind_rev =
+  Re.compile @@ Re.Posix.re {|^([$&*+\-/=>@^|.]|([a-zA-Z0-9_']*)%)(tel|dna)|}
+
+let infixRegex = Re.compile @@ Re.Posix.re {|[~?:!$&*+\-\/=><@^|%<.#]+$|}
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/prefix_parser.ml:17]: `infix_regex`

Comment (rgrinberg) [reply ocaml-lsp-server/src/prefix_parser.ml:17]: Also, the `regex` suffix isn't particularly useful. We already know what it is from the type

File: ocaml-lsp-server/bench/dune [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+  ocaml_lsp_server
+ core_unix.command_unix
+  merlin-lib.kernel
+
+  base
+  core
+  core_bench
+  )
+
+ (preprocess(pps ppx_jane ppx_bench))
```
Comment (rgrinberg) [inline ocaml-lsp-server/bench/dune:15]: You should run `make fmt`

File: ocaml-lsp-server/src/testing.ml [hunk 13]
```diff
@@ -0,0 +1,8 @@
+(**WARNING: This is for internal use in testing only *)
+
+module Compl=Compl
+module Document =Document
+module Import =Import
```
Comment (rgrinberg) [inline ocaml-lsp-server/src/testing.ml:5]: I don't see this module being used in the tests.

**Comment by rgrinberg** [comment]
> As for the benchmarks, I understand they may not make sense in this PR and you may want a different benchmarking lib etc. But I do think it makes sense to have something setup at some point. Even just for testing little things to ensure that new changes aren't causing perf regressions. I'm sure some people out there can always accurately guess the performance effect of a change just by looking at it, but I think microbenchmarks are a very good thing to have around.

I don't mind the benchmarking library you chose. It's just that there's so much overhead to getting completions that I doubt that running a regex is going to make any difference. Anyway, you can keep them.

> An example of this is that i find this server mostly very performant(kudos, the startup time is super impressive) but every now and again it gets into a weird state or i open a particular file that causes major slowdown. I'd love to find out why

Unfortunately most of the those performance problems are inherited from merlin itself.

File: ocaml-lsp-server/src/prefix_parser.ml (hunk 8 above)
Comment (rgrinberg) [reply ocaml-lsp-server/src/prefix_parser.ml:13]: Also, I think that if you used the combinators, constructing the "reverse" regex would be much simpler.

**Comment by faldor20** [comment]
Yeah I would have expected that Merlin would be the bottleneck in most cases. But I thought that I may as well ensure I'm not making it worse :). 
I think the recent release of.soemthing like bun in the JS ecosystem really highlights that great performance often comes from "sweating the small stuff" ,or at least paying attention to it.

Like it'd be great if Merlin and a robust benchmark suite that could identify the areas that could be improved. 

If you're happy I'll leave it as is.

File: ocaml-lsp-server/src/compl.ml [hunk 14]
```diff
@@ -26,8 +26,16 @@ let completion_kind kind : CompletionItemKind.t option =
...
+(* I should just rewrite all of this so that it uses a nice for loop. This
+   current soluction is a nice try but overall crap we need to be able to look
+   ahead and behind
+
+   Split it into name and infix name is obvious infix can be either a dot, a
+   label or an I could possibly do a regex based parser.
+
+   Name regex: ((\w)|\w.)*$ *)
+
+let prefix_of_position_old ~short_path source position =
```
Comment (jfeser) [inline ocaml-lsp-server/src/compl.ml:38]: Let's remove this function unless it's used elsewhere. We shouldn't keep it just for benchmarks.

File: ocaml-lsp-server/src/compl.ml [hunk 15]
```diff
@@ -95,8 +103,39 @@ let prefix_of_position ~short_path source position =
...
+  | text ->
+    let end_of_prefix =
+      let (`Offset index) = Msource.get_offset source position in
+      min (String.length text - 1) (index - 1)
+    in
+    let prefix_text =
+      let pos =
+        (*clamp the length of a line to process at 500 chars, this is just a
+          reasonable limit for regex performance*)
+        max 0 (end_of_prefix - 500)
```
Comment (jfeser) [inline ocaml-lsp-server/src/compl.ml:119]: I think we should avoid limits like this unless required.

**Comment by faldor20** [comment]
Okay, I have no idea why the formatter decided that it would be a good idea to format the result of that expect test, should be fixed now. Sorry this has been a bit of a mess, I'm actually traveling right now, so my work on this has been a bit erratic and distracted.

**Comment by faldor20** [comment]
Okay I actually have no idea why the build failed, I didn't change the doc_to_md file, at all.
I'll try to check more tomorrow when I at my laptop.

**Comment by faldor20** [comment]
Ahh I see, looks like I was hit with the issue fixed here https://github.com/ocaml/ocaml-lsp/pull/1204 could I get a rerun of CI @rgrinberg?

File: dune-project [hunk 16]
```diff
@@ -59,6 +59,7 @@ possible and does not make any assumptions about IO.
   (odoc-parser (and (>= 2.0.0) (< 2.3.0)))
   (ppx_expect (and (>= v0.15.0) :with-test))
   (ocamlformat (and :with-test (= 0.24.1)))
+  (ppx_bench (and :with-test (>= 0.16.0)))
```
Comment (rgrinberg) [inline dune-project:62]: Why do we need this for the tests?

**Comment by rgrinberg** [comment]
Thanks for the thorough PR and apologies for such a late response. I've merged it in https://github.com/ocaml/ocaml-lsp/pull/1212 with a minor tweaks.

The benchmarks had to be disabled in CI because there's no good way to report their results and they just slow things down with the unnecessary dependencies otherwise.

**Comment by faldor20** [comment]
All good, sorry I didn't quite have the time and expertise to finish that last little bit. Glad it got done :)
